content/chapters/compute/lab/content/user-level-threads.md

```diff
@@ -0,0 +1,87 @@
+
+The user can, however, implement further synchronization, by using the `yield()` call, or classic synchronization methods, like mutexes, barriers and condition variables.
+
+### Yielding
+
+As the scheduler is a cooperative one, each fiber can yield (or not), to allow another fiber to run.
+Follow the `support/user-level-threads/yield_launch.cc` implementation and run it.
+Note the `boost::fibers::launch::dispatch` parameter provided to the fiber constructor.
+It notifies the scheduler to start the fibre as soon as it is created.
+In order to explain the output, we must consider that the fibers are created by a **main fiber**, that is scheduled along with the others, in this case.
+
+Modify the launch parameter into `boost::fibers::launch::post`, compile and notice the differences.
+The `post` parameter notifies the scheduler not to start the fibers imediately, but rather place them into an execution queue.
+Their execution will start after the main fiber calls the `join()` function.
+
+### Barriers
+
+Follow the `support/user-level-threads/yield_barrier.cc` implementation.
+It uses a barrier to achieve the same result as the previos implementation, that used `post` as the launch parameter.
+
+## Interaction Between Threads and Fibers
+
+As we mentioned before, multiple fibers can run on the same thread, and a scheduler is implemented on each thread.
+By default, the scheduler is `round_robin`.
+It runs the fibers, in the order of their creation, until they yield or finish their work.
+If a fiber yields, it is placed at the back of the round-robin queue.
+Using this scheduler, each thread uses only its fibers;
+if one thread has more work to do than another, bad luck.
+
+But there are other scheduler implementations, such as `shared_work` and `work_stealing`.
+Follow the `support/user-level-threads/threads_and_fibers.cc` implementation.
+It creates multiple fibers and threads, and uses the `shared_work` scheduler to balance the workload between the threads.
+Each main fiber, from each thread, is suspended until all worker fibers have completed their work, using a condition variable.
+
+```cpp
+cnd_count.wait( lk, [](){ return 0 == fiber_count; } );
+```
+
+It also uses `thread local storage` and `fiber local storage` to store the ID of each thread / fiber.
+
+Now change the `shared_work` scheduler into the `work_stealing` one.
+It takes a parameter, the number of threads that will use that scheduler.
+
+Compile, rerun and note the differences.
+The `work_stealing` scheduler, as the name suggests, will "steal" fibers from other schedulers.
+So, if the `shared_work` scheduler tried to balance the available work between the available threads, the `work_stealing` one will focus on having as many threads as possible on 100% workload.
+Vary the number of threads and fibers, and the workload (maybe put each fibre to do some computational-intensive work), and observe the results.
```

Comment:
This is diving quite deeply into scheduling. It may not be useful here. I'd move this to the Arena and focus on the _usage_ of fibres here.

Reply:
It isn't much to do. It doesn't go into how the schedulers are implemented, just how they work. An arena exercise would be to implement your own scheduler, based on the `boost::fibers::algorithm` class.

Reply:
Yeah, I think you're right. Let's leave it here this semester and see how students react to it.